src/mds/MDBalancer.cc

```diff
@@ -385,94 +377,110 @@ double MDBalancer::try_match(mds_rank_t ex, double& maxex,
   return howmuch;
 }
 
-void MDBalancer::queue_split(CDir *dir)
+void MDBalancer::queue_split(const CDir *dir, bool fast)
 {
+  dout(10) << __func__ << " enqueuing " << *dir << dendl;
+
   assert(mds->mdsmap->allows_dirfrags());
-  split_queue.insert(dir->dirfrag());
-}
+  const dirfrag_t frag = dir->dirfrag();
 
-void MDBalancer::queue_merge(CDir *dir)
-{
-  merge_queue.insert(dir->dirfrag());
-}
+  auto callback = [this, frag](int r) {
+    if (split_pending.count(frag) == 0) {
+      // Someone beat me to it
+      return;
+    }
 
-void MDBalancer::do_fragmenting()
-{
-  if (split_queue.empty() && merge_queue.empty()) {
-    dout(20) << "do_fragmenting has nothing to do" << dendl;
-    return;
-  }
+    split_pending.erase(frag);
 
-  if (!split_queue.empty()) {
-    dout(10) << "do_fragmenting " << split_queue.size() << " dirs marked for possible splitting" << dendl;
+    CDir *split_dir = mds->mdcache->get_dirfrag(frag);
+    // FIXME maybe instead of this check we should ensure
```

Comment:
Yep.  I've gone with leaving things in split_pending even if they're no longer in the cache, because that list really just exists to avoid double-queueing the contexts that do the delayed split, and trying to cancel these contexts would be much harder than just checking if the dir is gone when we get called back.